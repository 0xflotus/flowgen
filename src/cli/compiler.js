/* @flow */
import { createSourceFile, ScriptTarget } from "typescript";
import fs from "fs";
import tsc from "typescript-compiler";

import namespaceManager from "../namespaceManager";
import { type Options, assignOptions, resetOptions } from "../options";
import { recursiveWalkTree } from "../parse";

const compile = (sourceFile): string => {
  const rootNode = recursiveWalkTree(sourceFile);

  const output = rootNode
    .getChildren()
    .map(child => {
      return child.print();
    })
    .join("");

  return output;
};

/**
 * Compiles typescript files
 */
export default {
  reset(options?: Options) {
    resetOptions();
    if (options) {
      assignOptions(options);
    }
    namespaceManager.reset();
  },

  compile,

  compileTest: (path: string, target: string): void => {
    tsc.compile(path, "--module commonjs -t ES6 --out " + target);
  },

  compileDefinitionString: (string: string, options?: Options): string => {
    resetOptions();
    if (options) {
      assignOptions(options);
    }
    namespaceManager.reset();

    return compile(
      createSourceFile("/dev/null", string, ScriptTarget.Latest, false),
    );
  },

  compileDefinitionFile: (path: string, options?: Options): string => {
    resetOptions();
    if (options) {
      assignOptions(options);
    }
    namespaceManager.reset();

    return compile(
      createSourceFile(
        path,
        fs.readFileSync(path).toString(),
        ScriptTarget.Latest,
        false,
      ),
    );
  },
};
