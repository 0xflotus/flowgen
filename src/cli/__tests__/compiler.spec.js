/* @flow */
import compiler from '../compiler';

it('should handle maybe & nullable type', () => {
  const result = compiler.compileDefinitionString('let a: string | null | undefined');

  expect(result).toMatchSnapshot();
});

it('should handle bounded polymorphism', () => {
  const ts = `
    function fooGood<T extends { x: number }>(obj: T): T {
      console.log(Math.abs(obj.x));
      return obj;
    }
  `;

  const result = compiler.compileDefinitionString(ts);

  expect(result).toMatchSnapshot();
});


it('should handle union strings', () => {
  const ts = `
    interface MyObj {
      state?: "APPROVED" | "REQUEST_CHANGES" | "COMMENT" | "PENDING"
    }
  `;

  const result = compiler.compileDefinitionString(ts);

  expect(result).toMatchSnapshot();
});
