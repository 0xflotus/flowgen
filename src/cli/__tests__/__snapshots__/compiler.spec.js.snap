// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should handle bounded polymorphism 1`] = `"declare function fooGood<T>(obj: T): T"`;

exports[`should handle maybe & nullable type 1`] = `"declare var a: string | null | void;"`;

exports[`should handle union strings 1`] = `
"declare interface MyObj {
state?: 'APPROVED' | 'REQUEST_CHANGES' | 'COMMENT' | 'PENDING'
} "
`;
